application/src/main/java/org/togetherjava/tjbot/commands/reminder/RemindRoutine.java

```diff
@@ -0,0 +1,143 @@
+
+import static org.togetherjava.tjbot.db.generated.Tables.PENDING_REMINDERS;
+
+/**
+ * Routine that processes and sends pending reminders.
+ * <p>
+ * Reminders can be set by using {@link RemindCommand}.
+ */
+public final class RemindRoutine implements Routine {
+    private static final Logger logger = LoggerFactory.getLogger(RemindRoutine.class);
+    private static final Color AMBIENT_COLOR = Color.decode("#F7F492");
+    private static final int SCHEDULE_INTERVAL_SECONDS = 30;
+    private final Database database;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param database the database that contains the pending reminders to send.
+     */
+    public RemindRoutine(@NotNull Database database) {
+        this.database = database;
+    }
+
+    @Override
+    public @NotNull Schedule createSchedule() {
+        return new Schedule(ScheduleMode.FIXED_RATE, 0, SCHEDULE_INTERVAL_SECONDS,
+                TimeUnit.SECONDS);
+    }
+
+    @Override
+    public void runRoutine(@NotNull JDA jda) {
+        Instant now = Instant.now();
+        database.write(context -> context.selectFrom(PENDING_REMINDERS)
+            .where(PENDING_REMINDERS.REMIND_AT.lessOrEqual(now))
+            .stream()
+            .forEach(pendingReminder -> {
+                sendReminder(jda, pendingReminder.getId(), pendingReminder.getChannelId(),
+                        pendingReminder.getAuthorId(), pendingReminder.getContent(),
+                        pendingReminder.getCreatedAt());
+
+                pendingReminder.delete();
+            }));
+    }
+
+    private static void sendReminder(@NotNull JDA jda, long id, long channelId, long authorId,
+            @NotNull CharSequence content, @NotNull TemporalAccessor createdAt) {
+        RestAction<ReminderRoute> route = computeReminderRoute(jda, channelId, authorId);
+        sendReminderViaRoute(route, id, content, createdAt);
+    }
+
+    private static RestAction<ReminderRoute> computeReminderRoute(@NotNull JDA jda, long channelId,
+            long authorId) {
+        // If guild channel can still be found, send there
+        TextChannel channel = jda.getTextChannelById(channelId);
+        if (channel != null) {
+            return createGuildReminderRoute(jda, authorId, channel);
+        }
+
+        // Otherwise, attempt to DM the user directly
+        return createDmReminderRoute(jda, authorId);
+    }
+
+    private static @NotNull RestAction<ReminderRoute> createGuildReminderRoute(@NotNull JDA jda,
+            long authorId, @NotNull TextChannel channel) {
+        return jda.retrieveUserById(authorId)
+            .onErrorMap(error -> null)
+            .map(author -> ReminderRoute.toPublic(channel, author));
+    }
+
+    private static @NotNull RestAction<ReminderRoute> createDmReminderRoute(@NotNull JDA jda,
+            long authorId) {
+        return jda.openPrivateChannelById(authorId)
+            .map(channel -> ReminderRoute.toPrivate(channel, channel.getUser()));
```

Comment:
Since we're just using PrivateChannel#getUser here, we can also just create a ReminderRouter#toPrivate(PrivateChannel) method? Allows us to use a method reference, and I love method references \:p

Reply:
Good spot

Reply:
Done. Please recheck, thanks.